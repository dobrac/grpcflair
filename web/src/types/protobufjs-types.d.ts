import protobuf from "protobufjs";
import { IFileDescriptorSet } from "protobufjs/ext/descriptor";

/**
 * Interface for ProtobufjsRootDescriptor for the root object support for gRPC reflection
 */
interface ProtobufjsRootDescriptor {
  toDescriptor(
    protoVersion: string,
  ): protobuf.Message<IFileDescriptorSet> & IFileDescriptorSet;
  fromDescriptor(
    descriptor:
      | protobuf.Message<IFileDescriptorSet>
      | IFileDescriptorSet
      | protobuf.Reader
      | Uint8Array,
  ): protobuf.Root;
}
