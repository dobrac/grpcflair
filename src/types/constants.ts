export const DEFAULT_URL = "/examples/helloworld.json";

export const DEFAULT_HOSTNAME = "http://localhost:4000";
