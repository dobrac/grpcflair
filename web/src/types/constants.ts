export const DEFAULT_URL = "/examples/helloworld.json";

export const DEFAULT_HOSTNAME = "http://localhost:8082";

export const FILTERED_NAMESPACES = ["google"];

export const AUTHORIZATION_METADATA_KEY = "authorization";
