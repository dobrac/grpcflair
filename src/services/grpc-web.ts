import protobuf from "protobufjs";
import {
  ClientReadableStream,
  GrpcWebClientBase,
  MethodDescriptor,
  MethodType,
} from "grpc-web";
import { GrpcWebFormat } from "@/types/grpc-web";

class DummyRPCType {
  constructor(...args: unknown[]) {}
}

export interface GrpcWebOptions {
  format?: GrpcWebFormat;
}

export function makeGrpcCall<MessageData extends object = object>(
  hostname: string,
  service: protobuf.Service,
  method: protobuf.Method,
  typeEncode: protobuf.Type,
  typeDecode: protobuf.Type,
  message: protobuf.Message,
  options?: GrpcWebOptions,
): Promise<protobuf.Message<MessageData>> {
  const client = new GrpcWebClientBase({ format: options?.format });

  const methodPath = `${hostname}/${
    // Replace starting dot "."
    service.fullName.replace(".", "")
  }/${method.name}`;

  return new Promise((resolve, reject) => {
    client.rpcCall(
      methodPath,
      // Ignored, using protobufjs directly
      {},
      {},
      new MethodDescriptor(
        method.name,
        MethodType.UNARY,
        // Ignored, using protobufjs directly
        DummyRPCType,
        // Ignored, using protobufjs directly
        DummyRPCType,
        () => {
          return typeEncode.encode(message).finish();
        },
        (bytes: Uint8Array) => {
          return typeDecode.decode(bytes);
        },
      ),
      (err, response: protobuf.Message<MessageData>) => {
        if (err) {
          reject(err);
        } else {
          resolve(response);
        }
      },
    );
  });
}

export function makeGrpcServerStreamingCall<
  MessageData extends object = object,
>(
  hostname: string,
  service: protobuf.Service,
  method: protobuf.Method,
  typeEncode: protobuf.Type,
  typeDecode: protobuf.Type,
  message: protobuf.Message,
  options?: GrpcWebOptions,
): ClientReadableStream<protobuf.Message<MessageData>> {
  if (options?.format !== GrpcWebFormat.TEXT) {
    throw new Error(
      `Only format ${GrpcWebFormat.TEXT} is supported for grpc-web server streaming`,
    );
  }

  const client = new GrpcWebClientBase({ format: options?.format });

  const methodPath = `${hostname}/${
    // Replace starting dot "."
    service.fullName.replace(".", "")
  }/${method.name}`;

  return client.serverStreaming<{}, protobuf.Message<MessageData>>(
    methodPath,
    // Ignored, using protobufjs directly
    {},
    {},
    new MethodDescriptor(
      method.name,
      MethodType.SERVER_STREAMING,
      // Ignored, using protobufjs directly
      DummyRPCType,
      // Ignored, using protobufjs directly
      DummyRPCType,
      () => {
        return typeEncode.encode(message).finish();
      },
      (bytes: Uint8Array) => {
        return typeDecode.decode(bytes);
      },
    ),
  );
}
