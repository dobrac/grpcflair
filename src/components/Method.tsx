import { useSourceContext } from "@/contexts/SourceContext";
import Loading from "@/components/Loading";
import { Button, Form } from "react-bootstrap";
import { makeGrpcCall, run } from "@/types/grpc-web";
import protobuf from "protobufjs";
import { useState } from "react";

export interface ServiceProps {
  service: protobuf.Service;
  method: protobuf.Method;
}

export default function Method({ service, method }: ServiceProps) {
  const { context } = useSourceContext();
  const [processing, setProcessing] = useState<boolean>(false);
  const [fieldValue, setFieldValue] = useState<string>("");
  const [response, setResponse] = useState<unknown | undefined>(undefined);

  const runTest = () => {
    if (!context) return;
    run(context);
  };

  if (!context) return <Loading />;

  const RequestType = service.lookupType(
    // TODO: Include full path?
    method.requestType,
  );
  const ResponseType = service.lookupType(
    // TODO: Include full path?
    method.responseType,
  );

  return (
    <div key={method.name}>
      <div className="d-flex gap-3 align-items-center my-2">
        <h4 className="m-0">{method.name}</h4>
        <Button size="sm" onClick={runTest}>
          runTest
        </Button>
      </div>
      <div className="card ps-4">
        <div>
          <h5>Request</h5>
          <div>Type: {method.requestType}</div>
          <div>Streaming: {(method.requestStream ?? false).toString()}</div>
          <div>
            {Object.values(RequestType.fields).map((field) => (
              <div key={field.name}>
                <div className="text-secondary fst-italic">{field.comment}</div>
                <div>{field.name}</div>
                <div>
                  <Form.Control
                    as="input"
                    onChange={(e) => setFieldValue(e.target.value)}
                    value={fieldValue}
                  />
                </div>
              </div>
            ))}
          </div>
          <div className="mt-2">
            <Button
              size="sm"
              disabled={processing}
              onClick={async () => {
                try {
                  setProcessing(true);

                  // TODO: Support request streaming
                  if (method.requestStream) {
                    throw new Error(
                      "Request streaming is not supported yet (client or bi-directional)",
                    );
                  }

                  const message = RequestType.create({
                    [Object.values(RequestType.fields)[0].name]: fieldValue,
                  });

                  const response = await makeGrpcCall(
                    service,
                    method,
                    RequestType,
                    ResponseType,
                    message,
                  );

                  setResponse(ResponseType.toObject(response, {}));
                } finally {
                  setProcessing(false);
                }
              }}
            >
              Call
            </Button>
          </div>
        </div>
        <div className="mt-3">
          <h5>Response</h5>
          <pre>
            {!processing && response
              ? JSON.stringify(response, null, 2)
              : "No response yet"}
          </pre>
          <div>Type: {method.responseType}</div>
          <div>Streaming: {(method.responseStream ?? false).toString()}</div>
        </div>
      </div>
    </div>
  );
}
