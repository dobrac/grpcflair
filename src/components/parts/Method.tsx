import { Badge, Button, Collapse, ProgressBar, Spinner } from "react-bootstrap";
import { makeGrpcCall, makeGrpcServerStreamingCall } from "@/types/grpc-web";
import protobuf from "protobufjs";
import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";
import YesNoIcon from "@/components/YesNoIcon";
import Type from "@/components/parts/Type";
import SyntaxHighlighter from "react-syntax-highlighter";
import docco from "react-syntax-highlighter/dist/esm/styles/hljs/docco";
import { serializeFieldDefaultValuesToJSON } from "@/types/protobufjs";
import InputFieldValue from "@/components/parts/inputfields/InputFieldValue";
import { RpcError } from "grpc-web";
import { useSourceContext } from "@/contexts/SourceContext";
import InputFieldName from "@/components/parts/inputfields/InputFieldName";
import PCancelable, { CancelError } from "p-cancelable";

enum RequestType {
  UNARY,
  CLIENT_STREAMING,
  SERVER_STREAMING,
  BIDIRECTIONAL_STREAMING,
}

function getRequestType(method: protobuf.Method): RequestType {
  if (method.requestStream && method.responseStream) {
    return RequestType.BIDIRECTIONAL_STREAMING;
  } else if (method.requestStream) {
    return RequestType.CLIENT_STREAMING;
  } else if (method.responseStream) {
    return RequestType.SERVER_STREAMING;
  } else {
    return RequestType.UNARY;
  }
}

function getColorFromMethod(method: protobuf.Method) {
  switch (getRequestType(method)) {
    case RequestType.BIDIRECTIONAL_STREAMING:
    case RequestType.CLIENT_STREAMING:
      return "danger";
    case RequestType.SERVER_STREAMING:
      return "dark";
    case RequestType.UNARY:
      return "success";
  }
}

function getMethodType(method: protobuf.Method) {
  switch (getRequestType(method)) {
    case RequestType.BIDIRECTIONAL_STREAMING:
      return "Bi-directional streaming";
    case RequestType.CLIENT_STREAMING:
      return "Client streaming";
    case RequestType.SERVER_STREAMING:
      return "Server streaming";
    case RequestType.UNARY:
      return "Unary";
  }
}

export interface ServiceProps {
  service: protobuf.Service;
  method: protobuf.Method;
}

export default function Method({ service, method }: ServiceProps) {
  method.resolve();

  const { hostname } = useSourceContext();
  const [open, setOpen] = useState(false);
  const [cancelFunction, setCancelFunction] = useState<(() => void) | null>(
    null,
  );
  const [error, setError] = useState<Error | null>(null);

  const RequestType = service.lookupType(
    // TODO: Include full path?
    method.requestType,
  );
  const ResponseType = service.lookupType(
    // TODO: Include full path?
    method.responseType,
  );

  const defaultRequestData = serializeFieldDefaultValuesToJSON(
    RequestType.fields,
  );
  const [requestData, setRequestData] =
    useState<Record<string, unknown>>(defaultRequestData);
  const [response, setResponse] = useState<unknown[]>([]);

  const processing = cancelFunction !== null;

  const handleUnaryRequest = async (message: protobuf.Message<{}>) => {
    const cancellablePromise = new PCancelable<protobuf.Message<{}>>(
      (resolve, reject, onCancel) => {
        const promise = makeGrpcCall(
          hostname,
          service,
          method,
          RequestType,
          ResponseType,
          message,
        );

        promise.then(resolve).catch(reject);
      },
    );

    const onCancel = () => {
      cancellablePromise.cancel();
    };
    setCancelFunction(() => onCancel);

    const response = await cancellablePromise;

    setResponse([ResponseType.toObject(response, {})]);
  };

  const handleServerStreaming = (message: protobuf.Message<{}>) => {
    return new Promise<void>((resolve, reject) => {
      const stream = makeGrpcServerStreamingCall(
        hostname,
        service,
        method,
        RequestType,
        ResponseType,
        message,
      );
      const onCancel = () => {
        stream.cancel();
        reject(new CancelError("Canceled"));
      };

      setCancelFunction(() => onCancel);

      stream.on("data", (data) => {
        const response = ResponseType.toObject(data, {});
        setResponse((responses) => [...responses, response]);
      });

      stream.on("error", (error) => {
        reject(error);
      });

      stream.on("end", () => {
        resolve();
      });
    });
  };

  const handleExecute = async () => {
    try {
      setCancelFunction(() => {});
      setResponse([]);
      setError(null);

      const message = RequestType.create(requestData);

      // TODO: Support request streaming
      if (method.requestStream) {
        throw new Error(
          "Request streaming is not supported yet (client or bi-directional)",
        );
      }

      if (method.responseStream) {
        await handleServerStreaming(message);
      }

      if (!method.requestStream && !method.responseStream) {
        await handleUnaryRequest(message);
      }
    } catch (e) {
      console.error(e);
      if (e instanceof Error) {
        setError(e);
      }
      if (e instanceof CancelError) {
        setError(new Error("Request canceled"));
      }
    } finally {
      setCancelFunction(null);
    }
  };

  const color = getColorFromMethod(method);

  return (
    <div key={method.name} className={"card bg-" + color + "-subtle"}>
      <button
        className="p-0 border-0 rounded-bottom-0 text-start btn hover-bg-darken py-2 px-3"
        onClick={() => setOpen((open) => !open)}
      >
        <div className="d-flex align-items-center gap-2">
          <Badge bg={color}>{getMethodType(method)}</Badge>
          <div className="fw-bold fs-6">
            {method.name}
            {!!method.requestStream && (
              <span className="fw-normal small text-secondary fst-italic ms-1">
                (Not supported yet)
              </span>
            )}
          </div>
          <div className="flex-grow-1 small text-secondary ms-1">
            {method.comment}
          </div>
          <div>
            {open ? (
              <FontAwesomeIcon icon={faChevronUp} />
            ) : (
              <FontAwesomeIcon icon={faChevronDown} />
            )}
          </div>
        </div>
      </button>
      <hr className="m-0" />
      <Collapse in={open}>
        <div>
          <div className="py-3 px-4 bg-light">
            <span className="fw-bolder">Parameters</span>
            {!!method.requestStream && (
              <span>
                streaming: <YesNoIcon value={true} className="ms-1" />
              </span>
            )}
          </div>
          <hr className="m-0" />
          <div className="py-2 px-4">
            <table className="w-100">
              <thead className="border-bottom border-secondary-subtle small">
                <tr>
                  <th>Name</th>
                  <th>Description</th>
                </tr>
              </thead>
              <tbody>
                {Object.values(RequestType.fields).map((field) => (
                  <tr key={field.name}>
                    <td className="align-top py-3">
                      <InputFieldName field={field} />
                    </td>
                    <td className="align-top py-3">
                      <div className="text-secondary">{field.comment}</div>
                      <InputFieldValue
                        field={field}
                        value={requestData[field.name]}
                        onChange={(value) => {
                          setRequestData((it) => {
                            return {
                              ...it,
                              [field.name]: value,
                            };
                          });
                        }}
                      />
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
            <div className="mt-2 d-grid gap-1">
              <Button size="sm" disabled={processing} onClick={handleExecute}>
                <Spinner
                  size="sm"
                  className={processing ? "visible" : "visually-hidden"}
                />{" "}
                Execute
              </Button>
              {processing && (
                <Button
                  size="sm"
                  variant="danger"
                  onClick={() => {
                    cancelFunction();
                  }}
                >
                  Cancel
                </Button>
              )}
            </div>
          </div>
          <hr className="m-0" />
          <div className="py-3 px-4 bg-light">
            <span className="fw-bolder">Responses</span>
            {!!method.responseStream && (
              <span>
                streaming: <YesNoIcon value={true} className="ms-1" />
              </span>
            )}
          </div>
          <hr className="m-0" />
          <div className="py-2 px-4 d-grid gap-4">
            <div className="d-grid gap-2">
              <div className="fw-bolder small">Request server</div>
              <div className="bg-dark text-light py-1 rounded-1 m-0 px-2">
                {hostname}
              </div>
            </div>
            <div className="d-grid gap-2">
              <div className="fw-bolder small">Server responses</div>
              {!response.length && !error && (
                <span className="small">No response yet</span>
              )}
              {processing && <ProgressBar animated now={100} />}
              {!!response.length && (
                <div className="d-grid gap-2">
                  {response.map((response, index) => (
                    <div key={index}>
                      <SyntaxHighlighter
                        language="json"
                        style={docco}
                        className="bg-dark text-light p-1 rounded-1 m-0"
                      >
                        {JSON.stringify(response, null, 2)}
                      </SyntaxHighlighter>
                    </div>
                  ))}
                </div>
              )}
              {!!error && (
                <div className="text-white bg-danger p-2 rounded-1">
                  {error instanceof RpcError && (
                    <>
                      <div>Error {error.code}</div>
                    </>
                  )}
                  <div>{error.message}</div>
                  {error instanceof RpcError && (
                    <>
                      <div>Metadata</div>
                      <SyntaxHighlighter
                        language="json"
                        style={docco}
                        className="bg-dark text-light p-1 rounded-1 m-0"
                      >
                        {JSON.stringify(error.metadata, null, 2)}
                      </SyntaxHighlighter>
                    </>
                  )}
                </div>
              )}
            </div>
            <div className="d-grid gap-2">
              <div className="fw-bolder small">Responses</div>
              <table className="w-100">
                <thead className="border-bottom border-secondary-subtle small">
                  <tr>
                    <th>Code</th>
                    <th>Description</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td className="align-top">0</td>
                    <td className="py-2">
                      <Type type={ResponseType} dark expanded={true} />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </Collapse>
    </div>
  );
}
