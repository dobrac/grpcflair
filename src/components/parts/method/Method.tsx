import { Badge, Collapse, ProgressBar } from "react-bootstrap";
import protobuf from "protobufjs";
import { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";
import YesNoIcon from "@/components/YesNoIcon";
import { getColorFromMethodType, getMethodType } from "@/services/protobufjs";
import { useSourceContext } from "@/contexts/SourceContext";
import ResponseError from "@/components/parts/method/response/ResponseError";
import ResponsesList from "@/components/parts/method/response/ResponsesList";
import ResponseExample from "@/components/parts/method/response/ResponseExample";
import { useMethodContext } from "@/contexts/MethodContext";
import JSONBlock from "@/components/JSONBlock";
import FormatSelector from "@/components/parts/method/request/FormatSelector";
import RequestFormExecution from "@/components/parts/method/request/RequestFormExecution";
import SectionBody from "./section/SectionBody";
import SectionHeader from "@/components/parts/method/section/SectionHeader";
import RequestForm from "@/components/parts/method/request/RequestForm";

export interface ServiceProps {
  service: protobuf.Service;
  method: protobuf.Method;
}

export default function Method({ service, method }: ServiceProps) {
  method.resolve();

  const { hostname } = useSourceContext();
  const { processing, response } = useMethodContext();

  const [open, setOpen] = useState(false);

  const responseType = method.resolvedResponseType;

  const color = getColorFromMethodType(method);

  return (
    <div key={method.name} className={"card bg-" + color + "-subtle"}>
      <button
        className="p-0 border-0 rounded-bottom-0 text-start btn hover-bg-darken py-2 px-3"
        onClick={() => setOpen((open) => !open)}
      >
        <div className="d-flex align-items-center gap-2">
          <Badge bg={color}>{getMethodType(method)}</Badge>
          <div className="fw-bold fs-6">
            {method.name}
            {!!method.requestStream && (
              <span className="fw-normal small text-secondary fst-italic ms-1">
                (Not supported yet)
              </span>
            )}
          </div>
          <div className="flex-grow-1 small text-secondary ms-1">
            {method.comment}
          </div>
          <div>
            {open ? (
              <FontAwesomeIcon icon={faChevronUp} />
            ) : (
              <FontAwesomeIcon icon={faChevronDown} />
            )}
          </div>
        </div>
      </button>
      <Collapse in={open}>
        <div>
          <SectionHeader>
            <div>
              <span className="fw-bolder">Parameters</span>
              {!!method.requestStream && (
                <span className="ms-1">
                  streaming: <YesNoIcon value={true} className="ms-1" />
                </span>
              )}
            </div>
          </SectionHeader>
          <SectionBody>
            <RequestForm method={method} />
            <RequestFormExecution service={service} method={method} />
          </SectionBody>
          <SectionHeader>
            <div>
              <span className="fw-bolder">Responses</span>
              {!!method.responseStream && (
                <span className="ms-1">
                  streaming: <YesNoIcon value={true} className="ms-1" />
                </span>
              )}
            </div>
            <div className="d-flex align-items-center gap-2">
              <span className="fw-bolder small text-nowrap">Content type</span>
              <FormatSelector />
            </div>
          </SectionHeader>
          <SectionBody>
            <div className="d-grid gap-4">
              <div className="d-grid gap-2">
                <div className="fw-bolder small">Request server</div>
                <JSONBlock>{hostname}</JSONBlock>
              </div>
              <div className="d-grid gap-2">
                <div className="fw-bolder small">Server responses</div>
                {!response && <span className="small">No response yet</span>}
                {processing && <ProgressBar animated now={100} />}
                <ResponsesList responses={response?.data ?? []} />
                <ResponseError error={response?.error} />
              </div>
              <div className="d-grid gap-2">
                <div className="fw-bolder small">Responses</div>
                <ResponseExample responseType={responseType} />
              </div>
            </div>
          </SectionBody>
        </div>
      </Collapse>
    </div>
  );
}
