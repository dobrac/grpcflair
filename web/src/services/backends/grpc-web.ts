import PCancelable, { CancelError } from "p-cancelable";
import { BackendDefinition, HandleRequest } from "./common";
import {
  makeGrpcCall,
  makeGrpcServerStreamingCall,
  UnaryResponse,
} from "@/services/grpc-web-client";
import { RequestType } from "@/services/protobufjs";

const handleUnaryRequestGrpcWeb: HandleRequest = async (
  hostname,
  service,
  method,
  requestType,
  responseType,
  message,
  metadata,
  functions,
  options,
) => {
  const cancellablePromise = new PCancelable<UnaryResponse>(
    (resolve, reject, onCancel) => {
      const promise = makeGrpcCall(
        hostname,
        service,
        method,
        requestType,
        responseType,
        message,
        {
          metadata,
          format: options?.format,
        },
      );

      promise.then(resolve).catch(reject);
    },
  );

  const onCancel = () => {
    cancellablePromise.cancel();
  };
  functions.setCancelFunction?.(() => onCancel);

  const response = await cancellablePromise;

  functions.setResponse?.((it) => ({
    ...it,
    headers: response.headers,
    trailers: response.trailers,
    data: response.data?.map((it) => responseType.toObject(it, {})),
  }));
};

const handleServerStreamingRequestGrpcWeb: HandleRequest = async (
  hostname,
  service,
  method,
  requestType,
  responseType,
  message,
  metadata,
  functions,
  options,
) => {
  return new Promise<void>((resolve, reject) => {
    const stream = makeGrpcServerStreamingCall(
      hostname,
      service,
      method,
      requestType,
      responseType,
      message,
      {
        metadata,
        format: options?.format,
      },
    );
    const onCancel = () => {
      stream.cancel();
      reject(new CancelError("Canceled"));
    };

    functions.setCancelFunction?.(() => onCancel);

    stream.on("data", (data) => {
      const response = responseType.toObject(data, {});
      functions.setResponse?.((it) => ({
        ...it,
        data: [...(it?.data ?? []), response],
      }));
    });

    stream.on("metadata", (metadata) => {
      functions.setResponse?.((it) => ({
        ...it,
        headers: metadata,
      }));
    });

    stream.on("status", (status) => {
      functions.setResponse?.((it) => ({
        ...it,
        trailers: status.metadata,
      }));
    });

    stream.on("error", (error) => {
      reject(error);
    });

    stream.on("end", () => {
      resolve();
    });
  });
};

export const grpcWebBackend: BackendDefinition = {
  [RequestType.UNARY]: handleUnaryRequestGrpcWeb,
  [RequestType.SERVER_STREAMING]: handleServerStreamingRequestGrpcWeb,
  [RequestType.CLIENT_STREAMING]: undefined,
  [RequestType.BIDIRECTIONAL_STREAMING]: undefined,
};
