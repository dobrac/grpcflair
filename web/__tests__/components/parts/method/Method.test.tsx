import "../../../../tests/mocks/grpc-web";
import { act, render } from "@testing-library/react";
import { context } from "../../../../tests/protobufjs-source";
import Method from "@/components/parts/method/Method";
import MethodContextProvider from "@/contexts/MethodContext";
import { DEFAULT_HOSTNAME } from "@/types/constants";
import {
  grpcWebMockHeaders,
  grpcWebMockResponseData,
  grpcWebMockTrailers,
} from "../../../../tests/mocks/grpc-web";

describe("Method", () => {
  const service = context.lookupService("helloworld.Greeter");
  service.resolve();

  const method = service.methods["TestInputTypes"];
  method.resolve();

  it("renders - collapsed", async () => {
    const renderedElement = await act(async () => {
      return render(
        <MethodContextProvider method={method}>
          <Method service={service} method={method} expanded={false} />
        </MethodContextProvider>,
      );
    });

    const element = renderedElement.container.firstChild;
    expect(element).toBeInTheDocument();

    expect(element).toHaveTextContent(method.name);

    if (method.comment) {
      expect(element).toHaveTextContent(method.comment);
    }
  });

  it("renders - expanded", async () => {
    const renderedElement = await act(async () => {
      return render(
        <MethodContextProvider method={method}>
          <Method service={service} method={method} expanded={true} />
        </MethodContextProvider>,
      );
    });

    const element = renderedElement.container.firstChild;
    expect(element).toBeInTheDocument();

    method.resolvedRequestType?.fieldsArray?.forEach((field) => {
      expect(element).toHaveTextContent(field.name);
    });
  });

  it("renders - execution", async () => {
    const method = service.methods["SayHello"];
    method.resolve();

    const renderedElement = await act(async () => {
      return render(
        <MethodContextProvider method={method}>
          <Method service={service} method={method} expanded={false} />
        </MethodContextProvider>,
      );
    });

    const element = await renderedElement.findByTestId("method-execute-button");
    expect(element).toBeInTheDocument();

    await act(async () => {
      element.click();
    });

    const responseRequestData = await renderedElement.findByTestId(
      "response-request-data",
    );
    expect(responseRequestData).toMatchSnapshot();

    const responseRequestHostname = await renderedElement.findByTestId(
      "response-request-hostname",
    );
    expect(responseRequestHostname).toHaveTextContent(DEFAULT_HOSTNAME);

    const responseResponses =
      await renderedElement.findByTestId("response-responses");
    expect(responseResponses).toHaveTextContent(grpcWebMockHeaders.header);
    expect(responseResponses).toHaveTextContent(
      grpcWebMockResponseData.message,
    );
    expect(responseResponses).toHaveTextContent(grpcWebMockTrailers.trailer);
  });
});
