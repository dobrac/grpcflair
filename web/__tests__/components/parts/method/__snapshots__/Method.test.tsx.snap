// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method renders - execution 1`] = `
<div
  class="d-grid gap-2"
  data-testid="response-request-data"
>
  <div
    class="fw-bolder small"
  >
    Request JSON
  </div>
  <div
    class="d-grid"
  >
    <div
      class="bg-body text-dark rounded-1 m-0 py-1 px-2 text-break rounded-bottom-0"
      data-testid="react-block"
    >
      <span
        class="small"
      >
        Metadata
      </span>
    </div>
    <pre
      class="bg-body text-dark rounded-1 m-0 py-1 px-2 text-break rounded-top-0"
      data-testid="json-block"
      style="display: block; overflow-x: auto; padding: 0.5em; color: rgb(0, 0, 0); background: rgb(248, 248, 255);"
    >
      <code
        class="language-json"
        style="white-space: pre-wrap;"
      >
        <span>
          <span>
            {}
          </span>
        </span>
      </code>
    </pre>
  </div>
  <div
    class="d-grid"
  >
    <div
      class="bg-body text-dark rounded-1 m-0 py-1 px-2 text-break rounded-bottom-0"
      data-testid="react-block"
    >
      <span
        class="small"
      >
        Request data
      </span>
    </div>
    <pre
      class="bg-body text-dark rounded-1 m-0 py-1 px-2 text-break "
      data-testid="json-block"
      style="display: block; overflow-x: auto; padding: 0.5em; color: rgb(0, 0, 0); background: rgb(248, 248, 255);"
    >
      <code
        class="language-json"
        style="white-space: pre-wrap;"
      >
        <span>
          <span>
            {}
          </span>
        </span>
      </code>
    </pre>
  </div>
</div>
`;
